src/encoding/bencode.rs

```diff
@@ -1,136 +1,125 @@
 #[derive(Debug)]
 pub enum Error {
-    FromUtf8,
     NonStandardChar(char)
 }
 
-pub trait Literal {
-    fn stringify(&self) -> Result<String, Error>;
-}
-
-// non primative bencode types
-pub type ListEntry = Box<dyn Literal>;
-pub type DictEntry = (ByteStr, Box<dyn Literal>);
 
-/*
-**  Bencode data structures
-**  (https://wiki.theory.org/index.php/BitTorrentSpecification#Bencoding)
-*/
+pub trait Encodable {
+    fn encode(&self) -> Result<String, Error>;
+}
 
 /*
 **  Byte strings are encoded as follows: <string length encoded in base ten ASCII>:<string data>
 **  Note that there is no constant beginning delimiter, and no ending delimiter.
 */
+impl Encodable for &str {
+    fn encode(&self) -> Result<String, Error> {
+        let mut bstring: Vec<char> = Vec::new();
+        for c in self.chars() {
+            if c as u32 > 127 {
+                return Err(Error::NonStandardChar(c))
+            }
 
-pub struct ByteStr {
-    content: Vec<u8>
-}
-
-impl Literal for ByteStr {
-    // create a String representation from ASCII sequence
-    fn stringify(&self) -> Result<String, Error> {
-        let arr = &self.content;
-        let ascii_result = String::from_utf8(arr.to_vec());
-        match ascii_result {
-            Ok(trailer) => Ok(format!("{}:{}", arr.len(), trailer)),
-            Err(_) => Err(Error::FromUtf8)
+            bstring.push(c);
         }
-    }
-}
 
-impl ByteStr {
-    // create a ByteStr from a string slice
-    pub fn try_from_str(ascii_128_str: &str) -> Result<Self, Error> {
-        let mut content: Vec<u8> = vec![];
 
-        // allocate chars to ASCII decimal vec
-        for c in ascii_128_str.chars() {
-            let decimal = c as usize;
-            // char incompatible with bencode
-            if decimal > 127 {
-                return Err(Error::NonStandardChar(c))
-            }
-            content.push(c as u8);
-        }
+        Ok(
+            format!("{}:{}", bstring.len(), bstring.iter().collect::<String>())
+        )
+    }
+}
 
-        Ok(ByteStr { content })
+impl Encodable for String {
+    fn encode(&self) -> Result<String, Error> {
+        Ok(Encodable::encode(&&self[..])?)
```

Comment:
- `&&self[..])?` Is this returning a new instance of Self by iterating its own chars?
- Why is the extra `&` needed, don't we already have a borrowed reference to self?

Reply:
The parameter `&self` is a `String` and it is being coerced into `&str` by taking the reference to the slice of itself.

Compiler complains without the extra `&` as it believes it is a `str` type, not `&str`.

Reply:
Ok. So is `..` in this case just short syntax for instantiating another `String` from `self`? Also need to clarify if by `str`, you mean `String` or literally a de-referenced string slice...